### ISSUE TITLE
Inconsistency between native and ARIA listboxes when implicit aria-selected is provided

### ISSUE THREAD

**Issue by joanmarie** [w3c/aria#1661]
According to the [spec](https://w3c.github.io/aria/#option):

> If a user agent provides an implicit `aria-selected` value for an `option`, the value SHOULD be true if the `option` has DOM focus or the `listbox` has DOM focus and the `option` is referenced by `aria-activedescendant`. Otherwise, if a user agent provides an implicit `aria-selected` value for an `option`, the value SHOULD be false.

This can lead to an inconsistency between native and ARIA listboxes. Consider the following:

```html
<select size="3">
<option>blue</option>
<option>red</option>
<option>green</option>
</select>
<button>ok</button>
```

If the user arrows down to the "red" `option` and then presses Tab, "red" is still selected even though the "ok" `button` is focused. Furthermore, and more importantly, a screen reader's review mode could be used to examine the `select` list and obtain the currently-selected value without ever changing focus. In the case of Linux, the AtkSelection interface implemented on that `select` element will return the accessible associated with the "red" `option`.

In contrast, we have this example:

```html
<div role="listbox" aria-activedescendant="red" tabindex="0">
<div role="option">blue</div>
<div role="option" id="red">red</div>
<div role="option">green</div>
</div>
<button>ok</button>
```

Assume the author provided keyboard navigation, the user arrowed to "red", the author updated the value of `aria-activedescendant` accordingly. Then the user pressed Tab to give focus to the "ok" `button`.

If the user agent provides an implicit `aria-selected` (which Chrome/Chromium does), the spec states that it should only report the selected item to ATs when the `listbox` or `option` is focused. Chrome/Chromium follows this. As a result,  a screen reader's review mode cannot be used to examine the `listbox` and obtain the currently-selected value without ever changing focus.

What is the rationale behind the language in the spec? Is there a benefit to not exposing the selected item when the `listbox` is no longer focused and introducing this inconsistency between native elements and ARIA?

**Comment by JAWS-test** [comment]
The problem with the implicit aria-selected is that it only works if the element has the focus. If it has no focus, there is no implicit aria-selected. How else could it be determined. At most by the element with tabindex=0. But that would be a new requirement that the selected option must have tabindex=0 - that does not exist at the moment. 

The easiest way to fix the problem would be to forbid implicit aria-selected and always require an element to be explicitly selected.

**Comment by JAWS-test** [comment]
https://w3c.github.io/aria/#aria-activedescendant: 

> Identifies the currently active element **when DOM focus is on** a composite widget, combobox, textbox, group, or application.

**Comment by scottaohara** [comment]
@joanmarie the quoted text [was added by this commit](https://github.com/w3c/aria/commit/488e3d467d1aeb2a4640a35047cf376a41855328), as a means to resolve #1052 dealing with ARIA's allowance of the checked state on options as well as selected (which was already an inconsistency between the native element and ARIA).

**Comment by smhigley** [comment]
Here's the PR link with comments and more context: https://github.com/w3c/aria/pull/1340

IIRC, we were trying to write the implicit selection/focus text to align with what was already the implemented browser behavior. And it already had precedence in the spec for treeitem implicit selection: https://github.com/w3c/aria/pull/1340/files#diff-0eb547304658805aad788d320f10bf1f292797b5e6d745a3bf617584da017051L12947 (line 12947, for anyone else who has issues with the link correctly jumping to the right line)

I think this issue predates that specific change, but it's now more explicitly a problem, and more widespread in the spec due to the PR 😅.

I think the main problem is that implicit selection on native controls does translate into an actual value/selection change for the form control. That would pretty much never be the case for an ARIA widget -- I can't think of a single custom web component I've seen that:
1. uses appropriate roles and also modifies its persisted internal selection state based on focus changes, but
2. doesn't update `aria-selected` to reflect that state.

So I think the crux of the issue is that implicit `aria-selected` isn't an internally persisted component state in practice, but implicit selection on native `<select>` is persisted. I can't see much point in UAs trying to maintain an implicit selected state for a control that doesn't have focus, since that state wouldn't actually be reflected by the control's own state. It'd probably be worse than doing nothing, since it would communicate to users that they did successfully select a value even though that is not what is authored.

I don't know what that means for the spec for implicit selection though 😅

**Comment by spectranaut** [comment]
I'm trying to follow this discussion but @smhigley and I wonder if I'm catching it all. I can't quite understand if there is a reason to allow "implicit selection" when the focus is on an ARIA listbox item in this scenario. By "implicit selection", I mean that according to the spec, the browser can say "selected=true" on a listbox item with focus.

From what you said, it sounds like this "implicit selection" is unnecessary because you can think of no ARIA widget that would not set aria-selected when the focus is on the listbox item. So at worse, I think this confusing spec language would lead to an very occasional bug if someone creates a poorly implemented ARIA listbox. Chrome, by the way, does do this "implicit selection", so if someone didn't set "aria-selected" on focus a UA would think an item was selected when it is not.

So I guess this all leads up to, for me, this question: What exactly is the scenario where we would want to allow the user agent to say "select=true" when the focus is on an item in an ARIA widget? If not listbox, then is there a different widget where implicit selection must be supported/allowed? If it's only for trees, for example, can we say you can only set the default value for aria-selected to be true on treeitems?

**Comment by aleventhal** [comment]
I'm not sure we should make any backwards incompatible changes. Even if none of us are aware of any listboxes or other widgets that rely on this behavior, and we think they always set aria-selected, doesn't mean they don't exist.

Could we fix this by:
- Saying this is a UA repair for no set selection in single selection containers
- Saying the other should or must still set aria-selected in a single selection container

Alternatively, we could still map the aria-activedescendant item to selected even when the container doesn't have focus.

**Comment by jnurthen** [comment]
https://www.w3.org/2022/01/06-aria-minutes.html#t12

**Comment by MelSumner** [comment]
> So I think the crux of the issue is that implicit aria-selected isn't an internally persisted component state in practice, but implicit selection on native <select> is persisted. I can't see much point in UAs trying to maintain an implicit selected state for a control that doesn't have focus, since that state wouldn't actually be reflected by the control's own state.  

Is there any point from the UA perspective? I agree with @smhigley here.

**Comment by carmacleod** [comment]
Starting with @aleventhal's alternative suggestion, as agreed to on the call:
> Alternatively, we [UA] could still map the aria-activedescendant item to selected even when the container doesn't have focus.

... and assuming that UAs can also map the item with tabindex="0" to selected in the case of a roving tabindex implementation

then we could change the [spec for option](https://w3c.github.io/aria/#option) slightly from:
> If a user agent provides an implicit aria-selected value for an option, the value SHOULD be true if the option has DOM focus or the listbox has DOM focus and the option is referenced by aria-activedescendant. Otherwise, if a user agent provides an implicit aria-selected value for an option, the value SHOULD be false.

to something like:
> If a user agent provides an implicit aria-selected value for an option, the value SHOULD be true if the option has DOM focus **or if it previously had focus and focus is no longer within the listbox**, or the listbox has DOM focus **or previously had focus** and the option is referenced by aria-activedescendant. Otherwise, if a user agent provides an implicit aria-selected value for an option, the value SHOULD be false.

Might need to also add the UA expectations to this section: https://w3c.github.io/aria/#managingfocus_useragents

Note that this does not address @smhigley's point 4, above. I'm not sure of a good way to address that, other than to say that authors MUST set aria-selected. But that would break backward compatibility, because [selection follows focus](https://w3c.github.io/aria-practices/#kbd_selection_follows_focus) has been allowed for a long time. I suppose we could say "authors SHOULD set aria-selected", and that might not break backward compatibility (?), or perhaps fuzzy words like, "For the best possible accessibility, use aria-selected". :)

Let me know if the simple wording change above (plus some tweaking of the UA managing focus section) is the direction we want to go in, and if so, I will write up a PR to that effect.

If we prefer to go the harsher, "authors SHOULD set aria-selected" route, let me know that instead.

**Comment by JAWS-test** [comment]
@carmacleod 

> if it previously had focus and focus is no longer within the listbox

I do not understand this addition. Shouldn't this apply regardless of whether the listbox already had the focus? It can also be pre-filled. Even then, it is important to be able to perceive the value. 

I think better would be:

> If a user agent provides an implicit aria-selected value for an option, the value SHOULD be true if the option has tabindex=0 (or tabindex with a value greater then 0), or the option is referenced by aria-activedescendant. Otherwise, if a user agent provides an implicit aria-selected value for an option, the value SHOULD be false.


The only problem we still have would be that an HTML select can also not have a selected value, while a listbox without aria-selected would always have a value (as soon as it has a tabindex=0 an option or an aria-activedescendant)

**Comment by carmacleod** [comment]
Agreed, @JAWS-test. I like your wording better because it covers more cases. Thanks!

**Comment by smhigley** [comment]
From ARIA WG discussion: we will leave the user agent text unchanged for the following reasons:
- One of the primary benefits of allowing AT to calculate implicit selection on focus is related to the historical behavior of announcing "not selected" instead of "selected". Implicit selection effectively means preventing every listbox option from announcing "not selected" unless specifically marked as such.
- Expanding the implicit selection beyond where it currently is defined increases the possibility of providing an incorrect and misleading selection announcement, especially given the wide variety of ways listboxes are coded in the wild.
- Deprecating the existing implicit selection on focus is too breaking

Instead, we'll add a normative requirement for authors to explicitly provide `aria-selected="true"` on the selected option for single-select, and `aria-selected="true/false"` on all options for multiselect.
